frontends/p4/tableKeyNames.cpp

```diff
@@ -57,6 +72,9 @@ class KeyNameGenerator : public Inspector {
         name.emplace(expression, l + "[" + r + "]");
     }
 
+    /** Compute a name annotation for @expression.  Eg. `@name("bits")` for
+     * `bits & 0x3` or `0x3 & bits`.
```

Comment:
Why do you say it is a P4-14 feature only?  The current p4c-bm2-ss compiler takes 'field_name & 0xff00 : exact;' as a key in a table with no error, and uses the mask feature in the bmv2 JSON file to represent it.  You can even put other expressions there like "field_name | 0xff00 : exact @name("foo");", but in that case the P4-16 compiler creates a temporary variable to hold the result of the expression, and then sends that as the table search key.

The P4-16 draft spec allows both of these things in the syntax.